src/record_ts.h

```diff
@@ -0,0 +1,873 @@
+    struct {
+        /// Represents the offset of that parameter for aggregate operations.
+        uint32_t offset;
+        /// Represents some literal data when aggregating literals.
+        uint64_t data;
+        /// The type size, when the actual type is not known. For example for
+        /// the aggregate operation.
+        int32_t type_size;
+    } extra;
+
+    ParamInfo() {}
+    ParamInfo(uint32_t index, VarType vtype) : slot(index), vtype(vtype) {}
+    ParamInfo(uint32_t index, uint32_t vtype)
+        : slot(index), vtype((VarType) vtype) {}
+};
+
+/**
+ * \brief Represents a frozen function recording. And can be used to replay it.
+ */
+struct Recording {
+    /// Weather this recording requires a dryrun, to discover the size of
+    /// certain operations.
+    bool requires_dry_run = false;
+
+    /// The variables used in this recording.
+    /// Each variable refers to an allocation.
+    /// If an allocation reuses a memory region, it is referred to by a separate
+    /// variable.
+    std::vector<RecordVariable> record_variables;
+
+    /// This vector maps the flat and deduplicated inputs to the frozen to their
+    /// variables in the \ref record_variables array.
+    std::vector<uint32_t> inputs;
+    /// This vector maps the flat outputs of the frozen function to their
+    /// recorded variables and how they have been accessed.
+    std::vector<ParamInfo> outputs;
+
+    /// Records the operations performed by this frozen function recording.
+    std::vector<Operation> operations;
+    /// Every operation refers to some number of variables, and encodes how they
+    /// are accessed. Instead of allocating a vector for each operation, \ref
+    /// Operation struct contains a pair that indexes into this vector.
+    std::vector<ParamInfo> dependencies;
+
+    /// The backend, which was used while recording.
+    JitBackend backend;
+
+    /// Replays the recording, given some inputs and fills the outputs with the
+    /// created variable indices.
+    /// Note that both \ref replay_input and \replay output have to have the
+    /// same size as the number of inputs and outputs with which the frozen
+    /// function was recorded.
+    int replay(const uint32_t *replay_input, uint32_t *outputs);
+
+    /// Counter, counting the number of kernels for debugging.
+    uint32_t n_kernels = 0;
+
+    /// This function is called after recording and checks that the recording is
+    /// valid i.e. that no variables where left uninitialized.
+    void validate();
+    /// Checks if all recorded kernels are still in the kernel cache.
+    /// This might occur when calling dr.kernel_cache_flush between recording
+    /// the function and replaying it.
+    bool check_kernel_cache();
+};
+
+/**
+ * \brief This struct is a wrapper arround a \ref ThreadState that records the
+ * operations performed with it.
+ */
+struct RecordThreadState : ThreadState {
+
+    RecordThreadState(ThreadState *internal) {
```

Comment:
Specify that this ThreadState does not take ownership of `internal`, so the `internal` ThreadState must stay alive as long as this new one is alive.